sqlite: Disallow writing from multiple `SQLiteBatch`s

achow101 opened the pull request:
The way that we have configured SQLite to run means that only one database transaction can be open at a time. Typically, each individual read and write operation will be its own transaction that is opened and committed automatically by SQLite. However, sometimes we want these operations to be batched into a multi-statement transaction, so `SQLiteBatch::TxnBegin`, `SQLiteBatch::TxnCommit`, and `SQLiteBatch::TxnAbort` are used to manage the transaction of the database.

However, once a db transaction is begun with one `SQLiteBatch`, any operations performed by another `SQLiteBatch` will also occur within the same transaction. Furthermore, those other `SQLiteBatch`s will not be expecting a transaction to be active, and will abort it once the `SQLiteBatch` is destructed. This is problematic as it will prevent some data from being written, and also cause the `SQLiteBatch` that opened the transaction in the first place to be in an unexpected state and throw an error.

To avoid this situation, we need to prevent the multiple batches from writing at the same time. To do so, I've implemented added a `CSemaphore` within `SQLiteDatabase` which will be used by any `SQLiteBatch` trying to do a write operation. `wait()` is called by `TxnBegin`, and at the beginning of `WriteKey`, `EraseKey`, and `ErasePrefix`. `post()` is called in `TxnCommit`, `TxnAbort` and at the end of `WriteKey`, `EraseKey`, and `ErasePrefix`. To avoid deadlocking on ` TxnBegin()` followed by a `WriteKey()`, `SQLiteBatch will now also track whether a transaction is in progress so that it knows whether to use the semaphore.

This issue is not a problem for BDB wallets since BDB uses WAL and provides transaction objects that must be used if an operation is to occur within a transaction. Specifically, we either pass a transaction pointer, or a nullptr, to all BDB operations, and this allows for concurrent transactions so it doesn't have this problem.

Fixes #29110

achow101 committed:
tests: Test for concurrent writes with db tx

There are occasions where a multi-statement tx is begun in one batch,
and a second batch is created which does a normal write (without a
multi-statement tx). These should not conflict with each other and all
of the data should end up being written to disk.

achow101 committed:
sqlite: Ensure that only one SQLiteBatch is writing to db at a time

A SQLiteBatch need to wait for any other batch to finish writing before
it can begin writing, otherwise db txn state may be incorrectly
modified. To enforce this, each SQLiteDatabase has a semaphore which
acts as a lock and is acquired by a batch when it begins a write, erase,
or a transaction, and is released by it when it is done.

To avoid deadlocking on itself for writing during a transaction,
SQLiteBatch also keeps track of whether it has begun a transaction.

furszy commented on src/wallet/sqlite.cpp:
```diff
@@ -606,10 +612,13 @@ std::unique_ptr<DatabaseCursor> SQLiteBatch::GetNewPrefixCursor(Span<const std::
+    m_database.m_sqlite_semaphore.wait();
     if (!m_database.m_db || sqlite3_get_autocommit(m_database.m_db) == 0) return false;
     int res = sqlite3_exec(m_database.m_db, "BEGIN TRANSACTION", nullptr, nullptr, nullptr);
     if (res != SQLITE_OK) {
         LogPrintf("SQLiteBatch: Failed to begin the transaction\n");
+    } else {
+        m_txn = true;
```
This can also cause a deadlock when the statement execution fails.

The semaphore is not released after a `TxnBegin` failure which will block forever if the software continues and executes a single write or another `TxnBegin` attempt. Which could happen as `TxnBegin` failure could be temporary.

achow101 replied:
Fixed

furszy reviewed (commented):
Found possible deadlock causes and crafted a test that exercises one of them: https://github.com/furszy/bitcoin-core/commit/20344b91a03259eb33652863700562f5ef198128. Feel free to pull it here.

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -606,10 +612,13 @@ std::unique_ptr<DatabaseCursor> SQLiteBatch::GetNewPrefixCursor(Span<const std::
 bool SQLiteBatch::TxnBegin()
 {
+    m_database.m_sqlite_semaphore.wait();
     if (!m_database.m_db || sqlite3_get_autocommit(m_database.m_db) == 0) return false;
```
Fixed

furszy committed:
test: wallet, coverage for concurrent db transactions

Verifying that a database handler can't commit/abort changes
occurring in a different database handler.

furszy reviewed (commented):
For the sake of completeness and to avoid potential issues, have expanded the test coverage to exercise the previously mentioned "concurrent database transactions deadlock" behavior. All yours: https://github.com/furszy/bitcoin-core/commit/a80cc56f0d1bdc82e9458f957db03c08620a6dd7. It verifies that the database handler cannot commit/abort changes occurring in a different database handler.

Also, while was working on this, discovered another deadlocking cause.. test https://github.com/furszy/bitcoin-core/commit/68e77b58e2284a92daff49a324fd229497da37ae. As it is a bdb-only issue, we could address it in another PR. But, it will mean that https://github.com/furszy/bitcoin-core/commit/a80cc56f0d1bdc82e9458f957db03c08620a6dd7 will need to be adapted to only run on sqlite (replace `TestDatabases()` for sqlite db creation call).

achow101 commented:
> Also, while was working on this, discovered another deadlocking cause.. test [furszy@68e77b5](https://github.com/furszy/bitcoin-core/commit/68e77b58e2284a92daff49a324fd229497da37ae). As it is a bdb-only issue, we could address it in another PR. But, it will mean that [furszy@a80cc56](https://github.com/furszy/bitcoin-core/commit/a80cc56f0d1bdc82e9458f957db03c08620a6dd7) will need to be adapted to only run on sqlite (replace `TestDatabases()` for sqlite db creation call).

Hmm, seems like BDB doesn't behave how I thought it did. It looks like we can't have concurrent transactions that try to operate on the same data, at least not without changing the isolation level, and I don't think it's worth the effort to figure that out when we're planning on removing BDB soon anyways. I've taken your test and updated it to just use SQLite.

achow101 commented on src/wallet/test/db_tests.cpp:
```diff
@@ -205,5 +205,31 @@ BOOST_AUTO_TEST_CASE(db_cursor_prefix_byte_test)
+        std::unique_ptr<DatabaseBatch> handler = Assert(database)->MakeBatch();
+
+        // Write original record
+        std::unique_ptr<DatabaseBatch> batch = database->MakeBatch();
```
Fixed

furszy commented on src/wallet/test/db_tests.cpp:
```diff
@@ -228,6 +228,48 @@ BOOST_AUTO_TEST_CASE(db_availability_after_write_error)
     }
 }
 
+BOOST_AUTO_TEST_CASE(concurrent_txn_dont_interfere)
+{
+    std::string key = "key";
+    std::string value = "value";
+    std::string value2 = "value_2";
+
+    DatabaseOptions options;
+    DatabaseStatus status;
+    bilingual_str error;
+    const auto& database = MakeSQLiteDatabase(m_path_root / "sqlite", options, status, error);
```
This will fail to compile without sqlite support. Need to scope the test under the `USE_SQLITE` compile flag. 
```diff
diff --git a/src/wallet/test/db_tests.cpp b/src/wallet/test/db_tests.cpp
--- a/src/wallet/test/db_tests.cpp	(revision f1b5b259f6d4759e7fc514b82b4096f7669e8325)
+++ b/src/wallet/test/db_tests.cpp	(date 1704374489489)
@@ -228,6 +228,7 @@
     }
 }
 
+#ifdef USE_SQLITE
 BOOST_AUTO_TEST_CASE(concurrent_txn_dont_interfere)
 {
     std::string key = "key";
@@ -269,6 +270,7 @@
     BOOST_CHECK(handler2->Read(key, read_value));
     BOOST_CHECK_EQUAL(read_value, value2);
 }
+#endif // USE_SQLITE
 
 
 BOOST_AUTO_TEST_SUITE_END()
```

achow101 replied:
Done

furszy reviewed (commented):
> I don't think this particular issue is a problem for us however. The fact that it deadlocks in this particular test case is probably a good thing since it means that BDB is handling locking internally. If there were two threads trying to read/write at the same time, BDB would force one thread to wait for the other, as this PR is doing for SQLite. That's the behavior that we want. It's only a deadlock (and test failure) since the test is trying to do the behavior of two threads within one thread.

The issue I envision is nested `WalletBatch` within the same thread. We need to be careful on not introducing an inner `WalletBatch` creation when there is an existing `WalletBatch` on any upper level.

Still, agree on not doing the changes here. Isn't related. I should revive #25297 after we finish #28574.

furszy reviewed (commented):
Another possible deadlock can arise when the txn abortion fails during the destruction of the batch object.
Essentially, if the txn abortion fails and the object is destructed, the semaphore remains locked indefinitely, leading to a deadlock in any subsequent write operation.

Made a test replicating the behavior; cherry-pick 5d2b79a65dcb10abdd349471c125595d677909ab and fbd59f89d3f5ed7bc026efa47d1f78c5cee04291 (from https://github.com/furszy/bitcoin-core/commits/sqlite-concurrent-writes/ )

achow101 commented:
> Another possible deadlock can arise when the txn abortion fails during the destruction of the batch object. Essentially, if the txn abortion fails and the object is destructed, the semaphore remains locked indefinitely, leading to a deadlock in any subsequent write operation.
> 
> Made a test replicating the behavior; cherry-pick [5d2b79a](https://github.com/bitcoin/bitcoin/commit/5d2b79a65dcb10abdd349471c125595d677909ab) and [fbd59f8](https://github.com/bitcoin/bitcoin/commit/fbd59f89d3f5ed7bc026efa47d1f78c5cee04291) (from https://github.com/furszy/bitcoin-core/commits/sqlite-concurrent-writes/ )

Changed `TxnAbort` and `TxnCommit` to always release the lock. Also pulled the test, and modified it to test `TxnCommit` and explicit `TxnAbort`

furszy reviewed (commented):
> Changed TxnAbort and TxnCommit to always release the lock.

This breaks the db txn isolation property.
The db txn is still active at the engine level, the lock shouldn't be released when the sqlite statement fails to execute (only when the abort/commit command executes successfully). Otherwise, other concurrent handlers are allowed to access db at the same time and, in the worst possible outcome, dump information to disk that should have been discarded.

Check this out 71170998. Made a test that exercises the behavior. Feel free to pull it.

Only the handler who originated the db txn should be able to commit and/or abort such txn, not other handlers. If other handlers were able to commit txns that they did not start, they could be saving incomplete information to disk and affecting other processes.

Aside from that, thanks for pulling all the tests commits. We are getting closer.

achow101 commented:
> The db txn is still active at the engine level, the lock shouldn't be released when the sqlite statement fails to execute (only when the abort/commit command executes successfully). Otherwise, other concurrent handlers are allowed to access db at the same time and, in the worst possible outcome, dump information to disk that should have been discarded.

If it fails to abort when the batch is destructed, then we're deadlocked. I don't see how this can be resolved without deadlocking and still preserving isolation.

furszy commented:
> If it fails to abort when the batch is destructed, then we're deadlocked. I don't see how this can be resolved without deadlocking and still preserving isolation.

Working on a solution. Will push it in few hours.

furszy reviewed (commented):
Done. Created #29253 to fix the root issue. I took some of my commits there.

With it, the fix for the problem https://github.com/bitcoin/bitcoin/pull/29112#pullrequestreview-1821862931 is straightforward. After resetting the database connection, we can release the lock, as closing the connection automatically rolls back the active transaction (see the PR). This ensures isolation across db handlers and will prevent the deadlocks here.

To show the end result, created a quick branch merging all the changes: This PR + #29253 + the test shared in [comment](https://github.com/bitcoin/bitcoin/pull/29112#pullrequestreview-1821862931) (71170998).  See https://github.com/furszy/bitcoin-core/tree/2024_pr29112_with_dangling_txn_fixes. All tests passing there :).

achow101 commented:
Reverted this to b316a81704fdd7fc5984ee8c189a937302b70c83 as #29253 deals with the problem of failing `TxnAbort()` and `TxnCommit()`.

ryanofsky reviewed (commented):
Code review fe617be10860cb5108399192ca65d85dc270622d. Looks good but I think there is still a race in SQLiteBatch::Close and a semaphore leak in SQLiteBatch::TxnBegin. Still reviewing, mostly just looked at the first commit so far.

achow101 commented on src/wallet/sqlite.h:
```diff
@@ -115,6 +117,10 @@ class SQLiteDatabase : public WalletDatabase
 
+    // Batches must acquire this semaphore on writing, and release when done writing.
+    // This ensures that only one batch is modifying the database at a time.
+    CSemaphore m_sqlite_semaphore;
```
Renamed

achow101 commented on src/wallet/sqlite.h:
```diff
@@ -58,6 +58,8 @@ class SQLiteBatch : public DatabaseBatch
     sqlite3_stmt* m_delete_stmt{nullptr};
     sqlite3_stmt* m_delete_prefix_stmt{nullptr};
 
+    bool m_txn{false};
```
Added a comment.

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -634,30 +645,41 @@ std::unique_ptr<DatabaseCursor> SQLiteBatch::GetNewPrefixCursor(Span<const std::
 {
+    if (m_txn) return false;
+    m_database.m_sqlite_semaphore.wait();
     if (!m_database.m_db || m_database.HasActiveTxn()) return false;
```
Done. Changed the `HasActiveTxn()` check to an assert.

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -634,30 +645,41 @@ std::unique_ptr<DatabaseCursor> SQLiteBatch::GetNewPrefixCursor(Span<const std::
 bool SQLiteBatch::TxnCommit()
 {
-    if (!m_database.HasActiveTxn()) return false;
+    if (!m_txn || !m_database.HasActiveTxn()) return false;
```
Done. Changed the `HasActiveTxn()` check to an assert.

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -634,30 +645,41 @@ std::unique_ptr<DatabaseCursor> SQLiteBatch::GetNewPrefixCursor(Span<const std::
 bool SQLiteBatch::TxnAbort()
 {
-    if (!m_database.HasActiveTxn()) return false;
+    if (!m_txn || !m_database.HasActiveTxn()) return false;
```
Done. Changed the `HasActiveTxn()` check to an assert.

furszy reviewed (commented):
Code reviewed, looking good. Left a last comment.

Also, have slightly modified my previous unit test to validate the changes (ba333a239b8aec4a7a6a79d11d4668730ee24803).
It is not really needed but feel free to take it if you like it (conceptually, it is similar to your test, the diff is that it is verifying the writes execution order and checks that when a batch finishes writing data to disk, another batch in a concurrent process can read the data immediately).

josibake commented:
Concept ACK

Started reviewing, looks good. Will continue review tomorrow. Might be worth updating the last part of the description with the insights from https://github.com/bitcoin/bitcoin/pull/29112#issuecomment-1874716791 ?

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -408,14 +408,16 @@ void SQLiteBatch::Close()
             // was opened will be rolled back and future transactions can succeed without committing old data.
             force_conn_refresh = true;
+            // If TxnAbort failed, the semaphore was not released, so release it here to avoid deadlocks on future writes.
+            m_database.m_write_semaphore.post();
```
Moved it.

furszy reviewed (commented):
ACK cfcb9b1ecf9be5267487713fa1e112ca09a1ae55

ryanofsky commented on src/wallet/test/db_tests.cpp:
```diff
@@ -279,8 +279,48 @@ BOOST_AUTO_TEST_CASE(txn_close_failure_dangling_txn)
 }
 
-#endif // USE_SQLITE
+BOOST_AUTO_TEST_CASE(concurrent_txn_dont_interfere)
```
In commit "test: wallet, coverage for concurrent db transactions" (cfcb9b1ecf9be5267487713fa1e112ca09a1ae55

Nothing about this test seems sqlite-specific. Seems like it would be nicer to call `TestDatabases` and be generic.

furszy replied:
This test was originally generic and it was moved to sqlite-only because of a bdb bug, see https://github.com/bitcoin/bitcoin/pull/29112#pullrequestreview-1795448339.

ryanofsky commented on test/functional/wallet_descriptor.py:
```diff
@@ -33,6 +36,41 @@ def skip_test_if_missing_module(self):
+            topup = thread.submit(wallet.keypoolrefill, newsize=1000)
+
+            # Then while the topup is running, we need to do something that will call
+            # ChainStateFlushed which will trigger a write to the db, hopefully at the
+            # same time that the topup still has an open db transaction.
```
In commit "tests: Test for concurrent writes with db tx" (548ecd11553bea28bc79e6f9840836283e9c4e99)

This seems fine, but you could probably increase the probability of this by calling keypoolrefill in a loop until gettxoutsetinfo returns instead of calling it once.

ryanofsky reviewed (approved):
Code review ACK cfcb9b1ecf9be5267487713fa1e112ca09a1ae55. This looks great and I think it is ready for merge. Just holding off because josibake seemed ready to review https://github.com/bitcoin/bitcoin/pull/29112#issuecomment-1930372190 and might have more feedback.

I did leave some comments but they are all minor and can be ignored. I was was also wondering about this in the PR description:

re: https://github.com/bitcoin/bitcoin/pull/29112#issue-2047565192

> To avoid this situation, we need to prevent the multiple batches from writing at the same time. Although SQLite does have facilities for doing that, they require the Write Ahead Log (WAL) mode, which is a feature that we explicitly chose to disable since it spread wallet data to multiple files. As such, we need to be handling the concurrency ourselves.

This doesn't seem accurate to me. As long as all the batch objects are sharing a `sqlite3* m_db` pointer and using the same connection to the database, a semaphore or mutex needs to be used to keep their transactions separate. This would be true whether the database is in WAL or journaling mode. Also if each batch object had it's own `sqlite3* m_db` connection, the semaphore should not be needed regardless of the database mode, IIUC.

achow101 commented on test/functional/wallet_descriptor.py:
```diff
@@ -33,6 +36,41 @@ def skip_test_if_missing_module(self):
 
+    def test_concurrent_writes(self):
+        self.log.info("Test sqlite concurrent writes are in the correct order")
+        self.restart_node(0, extra_args=["-unsafesqlitesync=0"])
```
The option's help text states `Set SQLite synchronous=OFF to disable waiting for the database to sync to disk. This is unsafe and can cause data loss and corruption. This option is only used by tests to improve their performance (default: false)`

achow101 commented:
> > To avoid this situation, we need to prevent the multiple batches from writing at the same time. Although SQLite does have facilities for doing that, they require the Write Ahead Log (WAL) mode, which is a feature that we explicitly chose to disable since it spread wallet data to multiple files. As such, we need to be handling the concurrency ourselves.
> 
> This doesn't seem accurate to me. As long as all the batch objects are sharing a `sqlite3* m_db` pointer and using the same connection to the database, a semaphore or mutex needs to be used to keep their transactions separate. This would be true whether the database is in WAL or journaling mode. Also if each batch object had it's own `sqlite3* m_db` connection, the semaphore should not be needed regardless of the database mode, IIUC.

Indeed, I misunderstood the documentation.
